server/src/main/java/org/elasticsearch/cluster/coordination/LeaderChecker.java

```diff
@@ -259,17 +262,28 @@ public String executor() {
                 });
         }
 
-        void leaderFailed() {
+        void leaderFailed(Exception e) {
             if (isClosed.compareAndSet(false, true)) {
-                transportService.getThreadPool().generic().execute(onLeaderFailure);
+                transportService.getThreadPool().generic().execute(new Runnable() {
+                    @Override
+                    public void run() {
+                        onLeaderFailure.accept(e);
+                    }
+
+                    @Override
+                    public String toString() {
+                        return "notification of leader failure: " + e.getMessage();
+                    }
+                });
             } else {
                 logger.trace("already closed, not failing leader");
             }
         }
 
         void handleDisconnectedNode(DiscoveryNode discoveryNode) {
             if (discoveryNode.equals(leader)) {
-                leaderFailed();
+                logger.debug("leader [{}] disconnected, restarting discovery", leader);
+                leaderFailed(new NodeNotConnectedException(discoveryNode, "disconnected"));
```

Comment:
why not `NodeDisconnectedException`?

Reply:
Typo - the keys are right next to each other. Fixed in f55db2a.